pkg/glusterfs/controllerserver.go

```diff
@@ -0,0 +1,377 @@
+package glusterfs
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	csi "github.com/container-storage-interface/spec/lib/go/csi/v0"
+	"github.com/gluster/glusterd2/pkg/api"
+	"github.com/golang/glog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+)
+
+const (
+	glusterDescAnn        = "GlusterFS-CSI"
+	glusterDescAnnValue   = "gluster.org/glusterfs-csi"
+	defaultVolumeSizeInMB = 20 * MB // minimum volume size
+)
+
+const (
+	_  = iota
+	KB = 1 << (10 * iota)
+	MB
+	GB
+	TB
+)
+
+type ControllerServer struct {
+	*GfDriver
+}
+
+func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
+
+	var glusterServer string
+	var bkpServers []string
+
+	if req == nil {
+		glog.Errorf("volume create request is nil")
+		return nil, status.Errorf(codes.InvalidArgument, "request cannot be empty")
+	}
+
+	if req.Name == "" {
+		return nil, status.Error(codes.InvalidArgument, "Name is a required field")
+	}
+	glog.V(1).Infof("creating volume with name ", req.Name)
+
+	if req.VolumeCapabilities == nil || len(req.VolumeCapabilities) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume capabilities is a required field")
+	}
+
+	// If capacity mentioned, pick that or use default size 20 MB
+	volSizeBytes := defaultVolumeSizeInMB
+	if req.GetCapacityRange() != nil {
+		volSizeBytes = int(req.GetCapacityRange().GetRequiredBytes())
+	}
+	volSizeMB := int((volSizeBytes + MB - 1) / MB)
+
+	// Get Volume name : TODO use the values from request
+	volumeName := req.Name
+	glusterVol := req.GetParameters()["glustervol"]
+	glusterServer = req.GetParameters()["glusterserver"]
+	glusterURL := req.GetParameters()["glusterurl"]
+	glusterURLPort := req.GetParameters()["glusterurlport"]
+	glusterUser := req.GetParameters()["glusteruser"]
+	glusterUserSecret := req.GetParameters()["glusterusersecret"]
+
+	glog.V(3).Infof("Request fields:[ %v %v %v %v %v %v]", glusterVol, glusterServer, glusterURL, glusterURLPort, glusterUser, glusterUserSecret)
+
+	// Get list of volumes in the TSP
+	volumeListResp, err := cs.client.Volumes("")
```

Comment:
Volume list can be skipped here. Try creating Volume, if fails with Exists error then get the volinfo and respond.

(Not for current PR, can be addressed later)

Reply:
sure, will get it in next pr.